source/simulator/core.cc

```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                initial_temperature_residual = system_rhs.block(introspection.block_indices.temperature).l2_norm();
+
+              const double temperature_residual = solve_advection(AdvectionField::temperature());
+
+              current_linearization_point.block(introspection.block_indices.temperature)
+                = solution.block(introspection.block_indices.temperature);
+              rebuild_stokes_matrix = true;
+              std::vector<double> composition_residual (parameters.n_compositional_fields,0);
+
+              for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
+                {
+                  assemble_advection_system (AdvectionField::composition(c));
+                  build_advection_preconditioner(AdvectionField::composition(c),
+                                                 C_preconditioner);
+
+                  if (nonlinear_iteration == 0)
+                    initial_composition_residual[c] = system_rhs.block(introspection.block_indices.compositional_fields[c]).l2_norm();
+
+                  composition_residual[c]
+                    = solve_advection(AdvectionField::composition(c));
+                }
+
+              // for consistency we update the current linearization point only after we have solved
+              // all fields, so that we use the same point in time for every field when solving
+              for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
+                current_linearization_point.block(introspection.block_indices.compositional_fields[c])
+                  = solution.block(introspection.block_indices.compositional_fields[c]);
+              LinearAlgebra::BlockVector global_backup_linearization_point = current_linearization_point;
+
+              if (use_picard == true && (max <= parameters.nonlinear_switch_tolerance ||
+                                         nonlinear_iteration >= parameters.max_pre_newton_nonlinear_iterations))
+                {
+                  use_picard = false;
+                  pcout << "Switching Newton derivatives on." << std::endl;
+
+                  bool use_said_method = false;
+                  if (!use_said_method)
+                    parameters.newton_residual = 0;
+                }
+
+              parameters.newton_theta = std::max(0.,(1-(parameters.newton_residual/parameters.switch_initial_newton_residual)));
+
+              if (nonlinear_iteration == 0)
+                {
+                  current_linearization_point.block(introspection.block_indices.velocities) = 0;
+
+                  rebuild_stokes_matrix = assemble_newton_stokes_system = true;
+                  rebuild_stokes_preconditioner = false;
+
+                  denormalize_pressure (this->last_pressure_normalization_adjustment,
+                                        current_linearization_point,
+                                        global_backup_linearization_point);
+
+                  compute_current_constraints ();
```

Comment:
Are those two calls necessary before setting the assemblers? It would look more reasonable, if they were placed right before the call where they are needed (I guess inside of `assemble_stokes_system`?)